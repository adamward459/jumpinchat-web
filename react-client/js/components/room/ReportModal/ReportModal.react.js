import React, { Component } from 'react';
import PropTypes from 'prop-types';
import Modal from 'react-modal';
import { setReportModal } from '../../../actions/ModalActions';
import { trackEvent } from '../../../utils/AnalyticsUtil';
import { sendReport } from '../../../utils/RoomAPI';

class ReportModal extends Component {
  constructor() {
    super();
    this.state = {
      error: null,
    };
    this.setReportModal = setReportModal;
    this.sendReport = sendReport;
    this.dismissModal = this.dismissModal.bind(this);
    this.submit = this.submit.bind(this);
  }

  dismissModal() {
    this.setState({ error: null });
    this.setReportModal(false);
    trackEvent('Report', 'Close report modal');
  }

  submit(e) {
    e.preventDefault();
    this.setState({ error: null });
    const {
      room,
      reporterId,
      targetId,
      messages,
    } = this.props;
    const reason = this.reason.value;
    const description = this.description.value;

    if (!reason) {
      return this.setState({ error: 'Select a reason' });
    }

    this.sendReport(room, reporterId, targetId, reason, description, messages);
    this.setReportModal(false);
  }

  render() {
    const {
      isOpen,
    } = this.props;

    const { error } = this.state;
    return (
      <Modal
        overlayClassName="modal"
        className="modal__Window"
        isOpen={isOpen}
        contentLabel="Room info modal"
      >
        <div className="modal__Header">
          Report user
        </div>
        <form onSubmit={this.submit}>
          <div className="modal__Body">
            <div className="form__InputGroup">
              <label
                htmlFor="reason"
                className="form__InputLabel"
              >
                Reason
              </label>
              <select
                id="reason"
                name="reason"
                className="input form__Input form__Input-inline form__Input-floating"
                ref={(e) => { this.reason = e; }}
              >
                <option value="" disabled selected>Please select a reason</option>
                <option value="harassment">Harassing or threatening me or a user</option>
                <option value="abuse">Abusive, defamatory or racist behaviour</option>
                <option value="nudity">Nudity</option>
                <option value="explicit">Explicit or offensive broadcast</option>
                <option value="other">Other (please describe below)</option>
              </select>
            </div>

            <div className="form__InputGroup">
              <label
                htmlFor="description"
                className="form__InputLabel"
              >
                More information (optional)
              </label>
              <textarea
                id="description"
                name="description"
                cols="30"
                rows="10"
                className="input form__Input form__Input-inline form__Input-floating form__Input-textarea"
                ref={(e) => { this.description = e; }}
              />
            </div>
            {error && (
              <span className="text text-red modal__Error">{error}</span>
            )}

            <p className="text-sub">
              Your visible chat log, as well as private conversations with the user being reported,
              will be sent as part of the report.
            </p>
            <p className="text-sub">
              Report users for violating
              the <a href="/terms" target="_blank">terms of service</a>.
              Please refrain from reporting users for actions that can be
              dealt with by room moderators, such as breaking room rules.
            </p>
            <p className="text-sub">
              See the <a href="/help/reporting" target="_blank">help page</a> for more
              information about reporting.
            </p>
          </div>
          <div className="modal__Footer">
            <button
              type="submit"
              className="button button-blue button-floating modal__Action"
            >
              Send
            </button>
            <button
              type="button"
              className="button button-default button-floating modal__Action"
              onClick={this.dismissModal}
            >
              Cancel
            </button>
          </div>
        </form>
      </Modal>
    );
  }
}

ReportModal.defaultProps = {
  targetId: null,
};

ReportModal.propTypes = {
  isOpen: PropTypes.bool.isRequired,
  room: PropTypes.string.isRequired,
  reporterId: PropTypes.string.isRequired,
  targetId: PropTypes.string,
  messages: PropTypes.arrayOf(PropTypes.shape({
    message: PropTypes.string.isRequired,
    userId: PropTypes.string,
    handle: PropTypes.string,
  })).isRequired,
};

export default ReportModal;
