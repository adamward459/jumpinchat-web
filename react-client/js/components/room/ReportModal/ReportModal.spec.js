import React from 'react';
import { shallow } from 'enzyme';
import ReportModal from './ReportModal.react';

describe('<ReportModal />', () => {
  let props;
  beforeEach(() => {
    props = {
      isOpen: false,
      reporterId: 'foo',
      room: 'room',
      messages: [{ message: 'bar' }],
    };
  });

  describe('dismissModal', () => {
    it('should clear error from state', () => {
      const wrapper = shallow(<ReportModal {...props} />);
      wrapper.instance().setState = jest.fn();
      wrapper.instance().dismissModal();
      expect(wrapper.instance().setState).toHaveBeenCalledWith({ error: null });
    });

    it('should set report modal state to false', () => {
      const wrapper = shallow(<ReportModal {...props} />);
      wrapper.instance().setReportModal = jest.fn();
      wrapper.instance().dismissModal();
      expect(wrapper.instance().setReportModal).toHaveBeenCalledWith(false);
    });
  });

  describe('submit', () => {
    let event;
    beforeEach(() => {
      event = { preventDefault: jest.fn() };
    });

    it('should set error state to null', () => {
      const wrapper = shallow(<ReportModal {...props} />);
      wrapper.instance().setState = jest.fn();
      wrapper.instance().reason = { value: false };
      wrapper.instance().description = { value: false };
      wrapper.instance().submit(event);
      expect(wrapper.instance().setState).toHaveBeenCalledWith({ error: null });
    });

    it('should set error if no reason selected', () => {
      const wrapper = shallow(<ReportModal {...props} />);
      wrapper.instance().setState = jest.fn();
      wrapper.instance().reason = { value: false };
      wrapper.instance().description = { value: false };
      wrapper.instance().submit(event);
      expect(wrapper.instance().setState)
        .toHaveBeenCalledWith({ error: 'Select a reason' });
    });

    it('should send a report', () => {
      props.targetId = 'target';
      const wrapper = shallow(<ReportModal {...props} />);
      wrapper.instance().sendReport = jest.fn();
      wrapper.instance().reason = { value: 'foo' };
      wrapper.instance().description = { value: 'bar' };
      wrapper.instance().submit(event);
      expect(wrapper.instance().sendReport)
        .toHaveBeenCalledWith('room', 'foo', 'target', 'foo', 'bar', [{ message: 'bar' }]);
    });
  });

  describe('render', () => {
    it('should show error message', () => {
      const wrapper = shallow(<ReportModal {...props} />);
      wrapper.setState({ error: 'err' });

      expect(wrapper.getElement()).toMatchSnapshot();
    });
  });
});
