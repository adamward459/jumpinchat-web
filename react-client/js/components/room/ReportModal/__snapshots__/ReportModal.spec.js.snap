// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ReportModal /> render should show error message 1`] = `
<Modal
  ariaHideApp={true}
  bodyOpenClassName="ReactModal__Body--open"
  className="modal__Window"
  closeTimeoutMS={0}
  contentLabel="Room info modal"
  isOpen={false}
  overlayClassName="modal"
  parentSelector={[Function]}
  portalClassName="ReactModalPortal"
  shouldCloseOnOverlayClick={true}
>
  <div
    className="modal__Header"
  >
    Report user
  </div>
  <form
    onSubmit={[Function]}
  >
    <div
      className="modal__Body"
    >
      <div
        className="form__InputGroup"
      >
        <label
          className="form__InputLabel"
          htmlFor="reason"
        >
          Reason
        </label>
        <select
          className="input form__Input form__Input-inline form__Input-floating"
          id="reason"
          name="reason"
        >
          <option
            disabled={true}
            selected={true}
            value=""
          >
            Please select a reason
          </option>
          <option
            value="harassment"
          >
            Harassing or threatening me or a user
          </option>
          <option
            value="abuse"
          >
            Abusive, defamatory or racist behaviour
          </option>
          <option
            value="nudity"
          >
            Nudity
          </option>
          <option
            value="explicit"
          >
            Explicit or offensive broadcast
          </option>
          <option
            value="other"
          >
            Other (please describe below)
          </option>
        </select>
      </div>
      <div
        className="form__InputGroup"
      >
        <label
          className="form__InputLabel"
          htmlFor="description"
        >
          More information (optional)
        </label>
        <textarea
          className="input form__Input form__Input-inline form__Input-floating form__Input-textarea"
          cols="30"
          id="description"
          name="description"
          rows="10"
        />
      </div>
      <span
        className="text text-red modal__Error"
      >
        err
      </span>
      <p
        className="text-sub"
      >
        Your visible chat log, as well as private conversations with the user being reported, will be sent as part of the report.
      </p>
      <p
        className="text-sub"
      >
        Report users for violating the 
        <a
          href="/terms"
          target="_blank"
        >
          terms of service
        </a>
        . Please refrain from reporting users for actions that can be dealt with by room moderators, such as breaking room rules.
      </p>
      <p
        className="text-sub"
      >
        See the 
        <a
          href="/help/reporting"
          target="_blank"
        >
          help page
        </a>
         for more information about reporting.
      </p>
    </div>
    <div
      className="modal__Footer"
    >
      <button
        className="button button-blue button-floating modal__Action"
        type="submit"
      >
        Send
      </button>
      <button
        className="button button-default button-floating modal__Action"
        onClick={[Function]}
        type="button"
      >
        Cancel
      </button>
    </div>
  </form>
</Modal>
`;
